proposals/b:zoe1/performActions.ts

```diff
@@ -0,0 +1,287 @@
+#!/usr/bin/env tsx
+
+import { execFileSync } from 'child_process'; // TODO: use execa
+import * as fsp from 'fs/promises';
+import * as path from 'path';
+import * as process from 'process';
+
+import { ZipReader } from '@endo/zip';
+import dbOpen from 'better-sqlite3';
+
+import assert from 'assert';
+import { makeAgd } from '../../upgrade-test-scripts/lib/agd-lib.js';
+import { agoric } from '../../upgrade-test-scripts/lib/cliHelper.js';
+import { voteLatestProposalAndWait } from '../../upgrade-test-scripts/lib/commonUpgradeHelpers.js';
+import { dbTool } from '../../upgrade-test-scripts/lib/vat-status.js';
+import { type WebCache } from '../../upgrade-test-scripts/lib/webAsset.js';
+import {
+  type ProposalInfo,
+  bundleDetail,
+  ensureISTForInstall,
+  flags,
+  getContractInfo,
+  loadedBundleIds,
+  txAbbr,
+} from './core-eval-support.js';
+
+// TODO move into core-eval-support
+const readSubmissions = async () => {
+  const files = await fsp.readdir('submission');
+  const names = files.filter(f => f.endsWith('.js')).map(f => f.slice(0, -3));
+  const buildAssets = {} as Record<string, ProposalInfo>;
+  for (const name of names) {
+    const evals = [{ permit: `${name}-permit.json`, script: `${name}.js` }];
+    const content = await fsp.readFile(`submission/${name}.js`, 'utf8');
+    const bundleIds = content.matchAll(/b1-[a-z0-9]+/g);
+    const bundles = Array.from(bundleIds).map(id => `${id}.json`);
+    buildAssets[name] = { evals, bundles };
+  }
+  return buildAssets;
+};
+
+/**
+ * URLs of assets, including bundle hashes (to be) agreed by BLD stakers
+ */
+const assetInfo = {
+  repo: {
+    name: 'zoe1',
+    description: 'first upgrade of Zoe vat',
+  },
+  buildAssets: await readSubmissions(),
+};
+
+const staticConfig = {
+  deposit: '10000000ubld', // 10 BLD
+  installer: 'gov1', // as in: agd keys show gov1
+  proposer: 'validator',
+  collateralPrice: 6, // conservatively low price. TODO: look up
+  swingstorePath: '~/.agoric/data/agoric/swingstore.sqlite',
+  title: assetInfo.repo.name,
+  description: assetInfo.repo.description,
+  buildInfo: Object.values(assetInfo.buildAssets),
+};
+
+const makeFakeWebCache = (base: string): WebCache => {
```

Comment:
Interfaces are great. I don't see why avoiding ambient authority makes them any more available. WebCache could have just as easily (more easily IMO) been implemented with ambient authority, just as this one is.

Reply:
My point is not that interfaces require no ambient authority, but that indiscriminate use of ambient authority in top-level functions leads to situations like [getNetworkConfig](https://github.com/Agoric/agoric-sdk/blob/master/packages/agoric-cli/src/lib/rpc.js#L37) where there's no way to virtualize the I/O that the function does, so none of the callers is reusable by replacing source of I/O access.

ocap design natually leads to objects shaped like `WebCache` that represent exactly the authority needed to do something and are fully virtualizable, rather than APIs that involve things like trading strings for I/O access.